### PR TITLE
Add support for `.framework`s to Carthage

### LINKED ISSUE #3174: Add Carthage frameworks to DependencyGraph
## Context 🕵️‍♀️

There is a case when Dependencies Graph is not quite accurate

## What 🌱

When we import `Carthage` dependency as a `binary`, `Carthage` `*.version` file is created without the `container` key and Dependecy graph relies on that key. So for now graph is just hiding this dependency with warning. Warning will be raised after [3164](https://github.com/tuist/tuist/pull/3164#issue-685323953) is merged.

## Proposal 🎉

So we need to add those `binary` dependencies to graph, but we also need to take into account various platforms if included in `Dependencies.swift`, such as `macOS`, `tvOS` and others. When imported as `binary`, `Carthage/Build` creates additional folder name by OS, for example. `iOS` or `Mac` so the path looks different, for example: `/Carthage/Build/iOS/GoogleMaps.framework`. 

The plan is to match the framework with the platform and give the `DependenciesGraph` it's content in the following format: `{dependencyName}.{dependencyPlatform}.framework`, for example: `GoogleMaps.iOS.framework`. That way we will have all the frameworks present in graph when requesting it by `tuist graph`

### PR DESCRIPTION
**Pull request by hisaac**
Resolves #3174

### Short description 📝

This change adds support for older `.framework` libraries to be added to the dependency graph when imported via Carthage.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
    - It passes linting, so I assume it's fine 🤷‍♂️
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
    - Haven't done any testing yet, beyond testing it with my individual project.
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
    - I'll do this once this pull request is out of draft status.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
    - I'll do this once this pull request is out of draft status.

### CODE REVIEW

**Comment by hisaac** [comment]
@laxmorek I'm re-opening this again, after moving the changes from my fork to the official tuist repo.

The changes I've made do seem to work fine with the project I'm migrating to Tuist, but obviously it will require more testing, and some changes to the unit tests. I'll start work on that tomorrow, but any advice on how to do that would be welcome!

File: Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift [hunk 1]
```diff
@@ -26,25 +28,37 @@ public final class CarthageGraphGenerator: CarthageGraphGenerating {
...
+                    let paths = platforms.map { $0.carthageDirectory }
+                    let frameworks = paths.map { path -> ProjectDescription.TargetDependency in
+                        pathString += path
+                        pathString += "/"
+                        pathString += product.name
+                        pathString += ".framework"
+                        return .framework(path: Path(pathString))
+                    }
+                    return frameworks
+                }
```
Comment (danyf90) [inline Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:58]: If more than one platform is specified in the Dependencies.swift, how is the correct platform selected at generation time?

Comment (hisaac) [reply Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:58]: I'm not yet familiar enough with the codebase to know the answer to that. Could you point me to the right direction on where the determination is happening in the code, or where I might add that?

Comment (danyf90) [reply Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:58]: The mapping is done in the `TargetDependency+ManifestMapper` file.

The problem is that with the current `DependenciesGraph` structure we don't have enough information to decide which framework should be picked.

It was briefly discussed [here](https://tuistapp.slack.com/archives/C01BYC3ND6U/p1629918751171900) by @laxmorek .

I would suggest discussing the desired solution in an RFC before going in solution mode with a PR

Comment (hisaac) [reply Sources/TuistDependencies/Carthage/Utils/CarthageGraphGenerator.swift:58]: Thanks @danyf90, I've seen that discussion and will likely implement one of the options discussed there. @laxmorek and I offered to guide me through working on this change, so I opened up this draft PR for he and I to discuss the work together.

**Comment by hisaac** [comment]
I've found a workaround for this in my project (just downloading the frameworks using Carthage directly, and pointing Tuist to the Carthage directory), so I won't be digging into this more at this time. @laxmorek, feel free to take this over if you'd like.

**Pull request closed without merging**